python/pyspark/mllib/clustering.py

```diff
@@ -38,13 +38,158 @@
+
+        :param point: the point to compute the cost to
+        """
+        return self.call("computeCost", _convert_to_vector(point))
+
+
+class BisectingKMeans:
+    """
+    A bisecting k-means algorithm based on the paper "A comparison of document clustering
+    techniques" by Steinbach, Karypis, and Kumar, with modification to fit Spark.
+    The algorithm starts from a single cluster that contains all points.
+    Iteratively it finds divisible clusters on the bottom level and bisects each of them using
+    k-means, until there are `k` leaf clusters in total or no leaf clusters are divisible.
+    The bisecting steps of clusters on the same level are grouped together to increase parallelism.
+    If bisecting all divisible clusters on the bottom level would result more than `k` leaf
+    clusters, larger clusters get higher priority.
+
+    Based on [[http://glaros.dtc.umn.edu/gkhome/fetch/papers/docclusterKDDTMW00.pdf
+    Steinbach, Karypis, and Kumar, A comparison of document clustering techniques,
+    KDD Workshop on Text Mining, 2000.]]
+    """
+    def __init__(self):
+        self.k = 4
+        self.maxIterations = 20
+        self.minDivisibleClusterSize = 1.0
+        self.seed = -1888008604  # classOf[BisectingKMeans].getName.##
+
+    def setK(self, k):
+        """
+        Set the number of leaf clusters.
+
+        :param k: the desired number of leaf clusters (default: 4). The actual number could be
+        smaller if there are no divisible leaf clusters.
+        """
+        self.k = k
+        return self
+
+    def getK(self):
+        """Return the desired number of leaf clusters."""
+        return self.k
+
+    def setMaxIterations(self, maxIterations):
+        """
+        Set the maximum number of iterations.
+
+        :param maxIterations: the max number of k-means iterations to split clusters (default: 20)
+        """
+        self.maxIterations = maxIterations
+        return self
+
+    def getMaxIterations(self):
+        """Return the maximum number of iterations."""
+        return self.maxIterations
+
+    def setMinDivisibleClusterSize(self, minDivisibleClusterSize):
+        """
+        Set the minimum divisible cluster size.
+
+        :param minDivisibleClusterSize: the minimum number of points (if >= 1.0) or the minimum
+        proportion of points (if < 1.0) of a divisible cluster (default: 1)
+        """
+        self.minDivisibleClusterSize = minDivisibleClusterSize
+        return self
+
+    def getMinDivisibleClusterSize(self):
+        """Return the min divisible cluster size."""
+        return minDivisibleClusterSize
+
+    def setSeed(self, seed):
+        """
+        Set the seed.
+
+        :param seed: a random seed (default: -1888008604)
+        """
+        self.seed = seed
+        return self
+
+    def getSeed(self):
+        """Return the random seed used."""
+        return self.seed
+
+    def run(self, rdd):
```

Comment:
I think we should follow the mllib rules of model training which use ```train``` rather than ```run```. Meanwhile, the ```set***``` and ```get***``` interfaces are unnecessary at mllib. You can refer [mllib.clustering.KMeans.train](https://github.com/apache/spark/blob/master/python/pyspark/mllib/clustering.py#L171).